skflow/monitors.py

```diff
@@ -0,0 +1,157 @@
+        if self.verbose and (self.steps % self.print_steps == 0):
+            self._set_training_summary()
+            print(self._summary_str)
+
+    def monitor_inducing_stop(self):
+        """Returns True if the monitor requests the model stop (e.g. for early stopping)"""
+        stop_now = (self.steps - self.min_loss_i >= self.early_stopping_rounds)
+        if stop_now:
+            sys.stderr.write("Stopping. Best step:\n step {} with loss {}\n"
+                             .format(self.min_loss_i, self.min_loss))
+        return stop_now
+
+    def create_val_feed_dict(self, inp, out):
+        """Validation requires access to TensorFlow placeholders. Not used in this Monitor"""
+        pass
+
+    def _set_epoch(self, feed_params_fn):
+        """Sets self.epoch from a function that generates a dictionary including this info"""
+        if feed_params_fn:
+            feed_params = feed_params_fn()
+            self.epoch = feed_params['epoch'] if 'epoch' in feed_params else None
+
+    def _set_training_summary(self):
+        """Returns the string to be written describing training progress"""
+        avg_train_loss = np.mean(self.print_train_loss_buffer)
+        self.print_train_loss_buffer = []
+        if self.epoch:
+            self._summary_str = ("Step #{step}, epoch #{epoch}, avg. train loss: {loss:.5f}"
+                                 .format(step=self.steps, loss=avg_train_loss,
+                                         epoch=self.epoch))
+        else:
+            self._summary_str = ("Step #{step}, avg. train loss: {loss:.5f}"
+                                 .format(step=self.global_step,
+                                         loss=avg_train_loss))
+        self._modify_summary_string()
+
+    def _modify_summary_string(self):
+        """Makes monitor specific changes to printed summary. Nothing interesting in BaseMonitor"""
+        pass
+
+
+class ValidationMonitor(BaseMonitor):
+    """Monitor that reports score for validation data and uses validation data for early stopping
+
+        val_X: Validation features
+        val_y: Validation labels
+        n_classes: Number of labels in output. 0 for regression
+        See BaseMonitor for arguments
+    """
+    def __init__(self, val_X, val_y, n_classes=0, *args, **kwargs):
+        super(ValidationMonitor, self).__init__()
+        self.val_feeder = setup_train_data_feeder(val_X, val_y, n_classes, -1)
+        self.print_val_loss_buffer = []
+        self.all_val_loss_buffer = []
+
+    def create_val_feed_dict(self, inp, out):
+        """Set tensorflow placeholders and create validation data feed"""
+        self.val_dict = self.val_feeder.get_feed_dict_fn(inp, out)()
+
+    def _set_last_loss_seen(self):
+        """Sets self.last_loss_seen to most recent validation loss
+
+        Also stores this value to appropriate buffers
+        """
+        [val_loss] = self.sess.run([self.loss_expression_tensor], feed_dict=self.val_dict)
```

Comment:
what if instead of passing inp/out/loss - pass the estimator itself, then it's possible to run predict and compute loss or any other function (for example if user wants to stop on accuracy) out of it.

not sure, what do you think about something like this?
```
def __init__(self, ..., n_classes=0, loss_fn=None)
    self.loss_fn = loss_fn
    ...

def update(self, estimator, ...):
     if self.loss_fn:
         self.loss_fn(estimator.predict(val_X), val_y)
    else:
         estimator.predict_loss(val_X, val_y) # this handles feed_dict and etc by a bit refactoring predict_ to support this case
```

where `loss_fn` if not None can be `sklearn.metrics.*`

I'm just a bit conscious that monitor has an interface that is very integrated with a concept of validation.